kernel/src/sched.rs

```diff
@@ -284,63 +309,33 @@ impl Kernel {
         }
     }
 
-    /// Main loop.
-    pub fn kernel_loop<P: Platform, C: Chip>(
-        &'static self,
-        platform: &P,
-        chip: &C,
-        ipc: Option<&ipc::IPC>,
-        _capability: &dyn capabilities::MainLoopCapability,
-    ) {
-        loop {
-            unsafe {
-                chip.service_pending_interrupts();
-                DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
-
-                for p in self.processes.iter() {
-                    p.map(|process| {
-                        self.do_process(platform, chip, process, ipc);
-                    });
-                    if chip.has_pending_interrupts()
-                        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                    {
-                        break;
-                    }
-                }
-
-                chip.atomic(|| {
-                    if !chip.has_pending_interrupts()
-                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                        && self.processes_blocked()
-                    {
-                        chip.sleep();
-                    }
-                });
-            };
-        }
-    }
-
-    unsafe fn do_process<P: Platform, C: Chip>(
+    unsafe fn do_process<P: Platform, C: Chip, S: SysTick>(
         &self,
         platform: &P,
         chip: &C,
+        systick: &S,
         process: &dyn process::ProcessType,
         ipc: Option<&crate::ipc::IPC>,
-    ) {
-        let systick = chip.systick();
+        timeslice_us: u32,
```

Comment:
```suggestion
        timeslice_us: Option<u32>,
```

Instead of having the cooperative (and other non-preempting schedulers) use the sentinel `0`, they should probably set `None` for this as they don't have "timeslices".